File: backend/services/activity_log_service.py

import logging
from datetime import datetime

from entities.activity_log import ActivityLog

logger = logging.getLogger(__name__)


class ActivityLogService:

    def __init__(self, supabase_client=None):
        self._supabase = supabase_client

    def record_log(self, user_id: str, activity: str, description: str) -> bool:
        if not self._supabase:
            logger.warning("Supabase tidak tersedia, log tidak dapat disimpan.")
            return False

        try:
            log_entry = ActivityLog(
                id=None,
                user_id=user_id,
                activity=activity,
                description=description,
                created_at=datetime.now(),
            )

            self._supabase.table("activity_logs").insert(
                {
                    "user_id": log_entry.user_id,
                    "activity": log_entry.activity,
                    "description": log_entry.description,
                }
            ).execute()

            logger.info(
                f"Log aktivitas dicatat — user: {user_id}, "
                f"aktivitas: '{activity}'"
            )
            return True

        except Exception as e:
            logger.error(f"Gagal menyimpan log aktivitas: {e}")
            return False

File: frontend/src/services/ActivityLogService.ts
import { supabase } from "@/lib/supabase";

export interface ActivityLog {
  id?: number;
  user_id: string;
  activity: string;
  description: string;
  created_at?: string;
}
class ActivityLogService {
  /**
   * Mencatat aktivitas user ke tabel activity_logs di Supabase.
   *
   * @param userId - UUID user yang melakukan aktivitas
   * @param activity - Nama singkat aktivitas (contoh: 'User Login')
   * @param description - Deskripsi detail aktivitas
   * @returns Promise<boolean> - true jika berhasil, false jika gagal
   */
  async recordLog(
    userId: string,
    activity: string,
    description: string
  ): Promise<boolean> {
    try {
      const { error } = await supabase.from("activity_logs").insert([
        {
          user_id: userId,
          activity,
          description,
        } satisfies Omit<ActivityLog, "id" | "created_at">,
      ]);

      if (error) {
        console.error("[ActivityLogService] Gagal mencatat log:", error.message);
        return false;
      }

      return true;
    } catch (err) {
      console.error("[ActivityLogService] Error tidak terduga:", err);
      return false;
    }
  }

  /**
   * Mengambil semua log aktivitas milik user tertentu dari Supabase.
   *
   * @param userId - UUID user
   * @returns Promise<ActivityLog[]> - Daftar log aktivitas
   */
  async getLogs(userId: string): Promise<ActivityLog[]> {
    try {
      const { data, error } = await supabase
        .from("activity_logs")
        .select("*")
        .eq("user_id", userId)
        .order("created_at", { ascending: false });

      if (error) throw error;
      return (data as ActivityLog[]) || [];
    } catch (err) {
      console.error("[ActivityLogService] Gagal mengambil log:", err);
      return [];
    }
  }
}

/** Singleton instance ActivityLogService untuk digunakan di seluruh aplikasi. */
export const activityLogService = new ActivityLogService();
